internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,558 @@
+
+	var err error
+	if p.nextCheckpoint != nil {
+		err = p.writeGetHeadersMsg(p.nextCheckpoint.Hash)
+	} else {
+		err = p.writeGetHeadersMsg(&zeroHash)
+	}
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) negotiateProtocol() error {
+	err := p.writeOurVersionMsg()
+	if err != nil {
+		return err
+	}
+	p.log.Info().Msgf("version sent to peer: %s", p)
+
+	err = p.readVerOrVerAck()
+	if err != nil {
+		return err
+	}
+
+	err = p.readVerOrVerAck()
+	if err != nil {
+		return err
+	}
+	p.log.Info().Msgf("received version and verack from peer: %s", p)
+
+	err = p.writeMessage(wire.NewMsgVerAck())
+	if err != nil {
+		return err
+	}
+
+	p.log.Info().Msgf("protocol negotiated successfully with peer: %s", p)
+	return nil
+}
+
+// PingHandler is a handler for sending ping messages to peers.
+// Must be run as a goroutine.
+func (p *Peer) pingHandler() {
+	p.wg.Add(1)
+	pingTicker := time.NewTicker(pingInterval)
+	defer pingTicker.Stop()
+
+out:
+	for {
+		select {
+		case <-pingTicker.C:
+			nonce, err := wire.RandomUint64()
+			if err != nil {
+				p.log.Error().Msgf("error generating nonce for ping msg, reason: %v", err)
+				continue
+			}
+			p.log.Info().Msgf("sending ping to peer %s with nonce: %d", p, nonce)
+			p.queueMessage(wire.NewMsgPing(nonce))
+
+		case <-p.quit:
+			break out
+		}
+	}
+
+	p.log.Info().Msgf("ping handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+// MsgHandler is a message handler for incoming messages.
+// Must be run as a goroutine.
+func (p *Peer) readMsgHandler() {
+	p.wg.Add(1)
+
+out:
+	for {
+		select {
+		case <-p.quit:
+			break out
+
+		default:
+			remoteMsg, _, err := wire.ReadMessage(p.conn, p.protocolVersion, p.chainParams.Net)
+			if err != nil {
+				if !p.quitting {
+					p.log.Error().Msgf("cannot read message from peer %s, reason: %v", p, err)
+				}
+				continue
+			}
+
+			switch msg := remoteMsg.(type) {
+			case *wire.MsgPing:
+				p.handlePingMsg(msg)
+			case *wire.MsgPong:
+				p.log.Info().Msgf("received pong from peer %s with nonce: %d", p, msg.Nonce)
+			case *wire.MsgHeaders:
+				p.handleHeadersMsg(msg)
+			case *wire.MsgInv:
+				p.handleInvMsg(msg)
+			default:
+				p.log.Info().Msgf("received msg of type: %T", msg)
+			}
+		}
+	}
+
+	p.log.Info().Msgf("read msg handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+func (p *Peer) writeMessage(msg wire.Message) error {
+	err := wire.WriteMessage(p.conn, msg, p.protocolVersion, p.chainParams.Net)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) writeRejectMessage(msg wire.Message, reason string) {
+	rejectMsg := wire.NewMsgReject(msg.Command(), wire.RejectObsolete, reason)
+	err := p.writeMessage(rejectMsg)
+	if err != nil {
+		p.log.Error().Msgf("could not write reject message to peer, reason: %v", err)
+	}
+}
+
+func (p *Peer) queueMessage(msg wire.Message) {
+	// running in goroutine here, because writing
+	// to channel is blocking until read
+	go func() {
+		p.msgChan <- msg
+	}()
+}
```

Comment:
As far as I understand your code - the `msgChan` is an *unbuffered* channel. So it will not act as a queue.
So the name of the method doesn't fit the implemented behaviour. 
Or I'm wrong 🙈 😄 


https://medium.com/@suryanshshrivastava_75738/unbuffered-channel-is-not-the-same-as-buffered-channel-of-size-1-3065c91344cc

Reply:
Switched to buffered channel. I decided to make a buffer with size 10 for now, because:

- we only send ping and getheaders msg, so 10 should be more than enough
- it's a buffer per peer, so every peer will have 10-msg size queue